backend/models/history.py

```diff
@@ -25,6 +25,8 @@ class OpsDBHistory(BaseModel):
     row_key = sa.Column(sa.String)
     original = sa.Column(JSONB)
     diff = sa.Column(JSONB)
+    changes = sa.Column(JSONB)
+    agreement_id = sa.Column(sa.Integer)
```

Comment:
Is there a way to do this without tying the `agreement_id` directly to the record? Thinking future, what if we wanted to query history for other specific models, like `CAN`? Would we need to embed the `can_id` as well? User History - `user_id`? etc. 

Is this data already accessible from the `original`, `diff`, or `changes` JSON blobs?

Reply:
I think it's possible without this column, but I haven't actually tried writing that alternative.  You need to query the even_details JSONB to find the agreement ID in different place depending on the class - ID for class_name in 'Agreement', 'ContractAgreement', 'GrantAgreement', ... or agreement_id for BLIs, etc.  I wasn't sure about creating an index for that later on.
I was thinking that Agreement was important enough to justify this, but not everything.  So far this Agreement-Details the only place in the UI where we collect all the history.  I wondered in Can or project might be.  User is probably different since it's already a column if you are looked at who made the change.

I can try the alternative using JSONB queries if that's preferred.

Reply:
WIP: here's a query using event_details JSONB to replace what's in agreement_history.py

```
stmt = select(OpsDBHistory).join(OpsDBHistory.created_by_user, isouter=True).add_columns(User)
stmt = stmt.where(
    and_(
        or_(
            and_(
                OpsDBHistory.event_details['id'].astext.cast(Integer) == id,
                OpsDBHistory.class_name.in_(class_names),
            ),
            OpsDBHistory.event_details['agreement_id'].astext.cast(Integer) == id
        ),
        OpsDBHistory.event_type.in_(
            [
                OpsDBHistoryType.NEW,
                OpsDBHistoryType.UPDATED,
                OpsDBHistoryType.DELETED,
            ]
        ),
    )
)
stmt = stmt.limit(limit)
if offset:
    stmt = stmt.offset(int(limit))
stmt = stmt.order_by(OpsDBHistory.created_on.desc())
```

If we want to go this route and eliminate agreement_id, we could consider eliminating row_key as well.  Although it is also being used to handle composite PKs, I'm not sure that any of the tables like that have Classes that will be logged in OpsDbHistory.

I am a little unsure about setting up indexes for this later on.  Before it was just three columns, now it's using a mix of stuff inside JSONB with other columns.

Reply:
agreement_id removed and JSONB query used instead